src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -127,18 +128,22 @@ static ucs_status_t uct_cuda_ipc_is_peer_accessible(uct_cuda_ipc_component_t *md
     char* accessible;
     CUdeviceptr d_mapped;
 
+    pthread_mutex_lock(&uct_cuda_ipc_map_lock);
```

Comment:
That's correct. It's not related to this PR but it is still needed. Hope it's ok to have this as part of this PR. 

I've moved the lock into the MD object.

Reply:
maybe make it spinlock and reduce lock scope to only updating the array (move cuda API calls outside of lock section)

Reply:
Switched to spinlock and reduced scope. If the array element is in init state (`0xFF`), then we want a thread to query CUDA API and update this. This write needs to be in the lock section. Also, we want at most one thread to call cuIpcOpenMemHandle and update the array element value from init to reachable or unreachable. We can't have have multiple threads calling cuIpcOpenMemHandle because that can lead to "already opened" error. With these two constraints, I couldn't think of a way to move cuda API out of the lock section.

Reply:
since we probably don't care if more than one thread would update "accessible" at the same time, we don't even need a lock.
since single-value memory read/write is atomic (e.g will get one of the values previously written, and not some garbage), can do it like this:
```
ipc_md {
    ucs_ternary_value_t accessible[];
};
...
uct_cuda_ipc_is_peer_accessible()
{
ucs_ternary_value_t is_accessible = *accessible;
if (is_accessible == UCS_TRY) {
   cuIpc(...)
   is_accessible = (result == ...) ? UCS_YES : UCS_NO;
  *accessible = is_accessible;
}
return is_accessible==UCS_YES;

```

Reply:
> since we probably don't care if more than one thread would update "accessible" at the same time, we don't even need a lock.

Agreed. We don't need lock on accessible array.

We still need no more than one thread call cuIpcOpenMemhandle. The following code doesn't prevent that.
```
uct_cuda_ipc_is_peer_accessible() {
...
ucs_ternary_value_t is_accessible = *accessible;
if (is_accessible == UCS_TRY) {
...
}
```

Two threads could end up calling OpenMemHandle on the same memory handle if they both see is_accessible as `UCS_TRY`, right? We may need all of the following executed in atomic fashion to disallow that:
```
if (is_accessible == UCS_TRY) {
   cuIpc(...)
   is_accessible = (result == ...) ? UCS_YES : UCS_NO;
  *accessible = is_accessible;
}
```
Don't we need a lock for this? Lmk if I've misunderstood something here.

Reply:
Why 2 threads opening the same handle is a problem?
seems we treat CUDA_ERROR_ALREADY_MAPPED staus as success

Reply:
@yosefe you're right. I ignored that aspect. This reminded me of another motivation for adding the lock which was to avoid opening the memory handle multiple times from the perspective of each thread because it can be in the order of ms. Even though opening the memory handle is expensive it's probably still ok to not have a lock because we would have accessible element set to UCS_YES soon that not many threads will end up opening the handle anyway. I'll make changes based on your code suggestion above.